cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+    and Bravyi's CH-form. The tableaux keeps track of the stabilizer operations, while the
+    CH-form allows access to the full wavefunction (including phase).
+
+    Gates and measurements are applied to each representation in O(n^2) time. 
+    """
+    def __init__(self,qubit_map,initial_state=0):
+        self.qubit_map = qubit_map
+        self.n = len(qubit_map)
+
+        self.tableau = CliffordTableau(self.n,initial_state)
+        self.CH_form = CH_Form(self.n,initial_state)
+            
+    def copy(self):
+        state = CliffordState(self.qubit_map)
+        state.tableau = self.tableau.copy()
+        state.CH_form = self.CH_form.copy()
+
+        return state
+            
+    def __repr__(self):
+        return self.CH_form.__repr__()
+
+    def to_numpy(self):
+        return self.CH_form.to_numpy()
+
+    # def print_stabilizers(self):
+    #     print(self.tableau.__str__())
+
+    def stabilizers(self):
+        return self.tableau.stabilizers()
+
+    def wave_function(self):
+        return self.CH_form.wave_function()
+            
+    def apply_unitary(self, op: ops.Operation):
+        if op.gate == cirq.CNOT:
+            self.tableau._CNOT(self.qubit_map[op.qubits[0]],self.qubit_map[op.qubits[1]]) 
+            self.CH_form._CNOT(self.qubit_map[op.qubits[0]],self.qubit_map[op.qubits[1]]) 
+        elif op.gate == cirq.CZ:
+            self.tableau._CZ(self.qubit_map[op.qubits[0]],self.qubit_map[op.qubits[1]]) 
+            self.CH_form._CZ(self.qubit_map[op.qubits[0]],self.qubit_map[op.qubits[1]]) 
+        elif op.gate == cirq.Z:
+            self.tableau._Z(self.qubit_map[op.qubits[0]])
+            self.CH_form._Z(self.qubit_map[op.qubits[0]])
+        elif op.gate == cirq.X:
+            self.tableau._X(self.qubit_map[op.qubits[0]])
+            self.CH_form._X(self.qubit_map[op.qubits[0]])
+        elif op.gate == cirq.Y:
+            self.tableau._Y(self.qubit_map[op.qubits[0]])
+            self.CH_form._Y(self.qubit_map[op.qubits[0]])
+        elif op.gate == cirq.S:
+            self.tableau._S(self.qubit_map[op.qubits[0]])
+            self.CH_form._S(self.qubit_map[op.qubits[0]])
+        elif op.gate == cirq.H:
+            self.tableau._H(self.qubit_map[op.qubits[0]])
+            self.CH_form._H(self.qubit_map[op.qubits[0]])
+        else:
+            raise ValueError('%s cannot be run with Clifford simulator' % str(op.gate))
+            
+    def perform_measurement(self, qubits: List[ops.Qid], collapse_wavefunction = True):
+        results = []
+        
+        if collapse_wavefunction:
+            state = self
+        else:
+            state = self.copy()
+        
+        for qubit in qubits:
+            result = state.tableau._measure(self.qubit_map[qubit])
+            state.CH_form.project_Z(self.qubit_map[qubit],result)
+            results.append(result)
+        
+        return results
+
+class CliffordTableau():
```

Comment:
Split this class out into a separate file.

Reply:
Done.